pkg/kubectl/resource/categories.go

```diff
@@ -36,6 +36,58 @@ func (e SimpleCategoryExpander) Expand(category string) ([]schema.GroupResource,
 	return ret, ok
 }
 
+type discoveryCategoryExpander struct {
+	fallbackExpander CategoryExpander
+	discoveryClient  discovery.DiscoveryInterface
+}
+
+// NewDiscoveryCategoryExpander returns a category expander that makes use of the "categories" fields from
+// the API, found through the discovery client. In case of any error or no category found (which likely
+// means we're at a cluster prior to categories support, fallback to the expander provided.
+func NewDiscoveryCategoryExpander(fallbackExpander CategoryExpander, client discovery.DiscoveryInterface) (discoveryCategoryExpander, error) {
+	if client == nil {
+		return discoveryCategoryExpander{}, errors.New("Please provide discovery client to shortcut expander")
+	}
+	return discoveryCategoryExpander{fallbackExpander: fallbackExpander, discoveryClient: client}, nil
+}
+
+func (e discoveryCategoryExpander) Expand(category string) ([]schema.GroupResource, bool) {
+	apiResourceLists, err := e.discoveryClient.ServerResources()
+	if err != nil {
```

Comment:
I would run this check differently.  The `ServerResources` signature can return an error *and* a valid (but partial) list of resources.  If you have any resources, I would continue. If you have zero resources, I would fallback.

Reply:
@deads2k Should I check both (fallback if zero resources **and** error)?

Reply:
Only fallback if there are zero resources, otherwise continue.